### DISCUSSION TITLE
Worksheet not working

### DISCUSSION THREAD

**Discussion by w-r-z-k** [scalameta/nvim-metals#294] [General]
It seems that the workspace (ed: I meant *worksheet*) isn't working.  I don't think the settings for nvim-metals recognizes *.workspace.sc as a workspace.

If I install nvim from scratch and setup nvim-metals, it works.  But what I have done is install the [lunarvim IDE layer](https://github.com/LunarVim/LunarVim) on nvim and then [activated nvim-metals via that mechanism](https://www.lunarvim.org/languages/scala.html).  So I suspect that there's some setting in lunarvim that is overriding the setting in nvim-metals.  If I knew what that setting was perhaps I could make it work.

For context: I'm neither an nvim configuration expert nor that conversant with lua (which is why I'm using lunarvim to spare me a very large amount of IDE configuration work).  Just the stock nvim install with nvim-metals works but leaves me with a good scala IDE experience but not for other IDE functions overall.  lunarvim adds a good experience.  But there's some gaps with using scala which nvim-metals addresses.  So I'm stuck in between and looking for the best of both worlds.

**Comment by ckipp01** [comment]
Hey @w-r-z-k I'm not very familiar with Lunarvim, so I'm not sure how much help I'll be especially if everything is working with `nvin-metals` directly, but not when working with Lunarvim.  A couple questions pop into my mind as I'm reading.

> I don't think the settings for nvim-metals recognizes *.workspace.sc as a workspace.

Do you mean `*worksheet.sc`? And if so what is mean by workspace here? Do you mean workspace root, and that Metals doesn't start?

Looking at the linked Scala section I'm not 100% sure what these things are doing. Primarily:

```
metals_config.on_attach = function()
          require("lsp").common_on_attach()
end
```
Any idea what the `common_on_attatch()` is doing? `nvim-metals` works completely independant of nvim-lspconfig,  so I'm hoping this attatch isn't reliant on that. I'm going to assume this just setting up mappings. The other thing I notice is that it's just calling the `initialize_or_attach`, which should be fine, but how is this triggered. With Lunarvim, is it just based on filetype?

**Reply by w-r-z-k** [reply]
Sorry, I meant worksheet.sc.  I'm not sure how LunarVim works.  It's mean to provide a large set of configurations for a Vim IDE as a black box; but on the other hand if something isn't working there's no way to unravel how it works. Thanks.

**Reply by ckipp01** [reply]
I'd recommend creating an issue in their repo about worksheets not working. If needed I can provide insight on what is needed for it to work there, and maybe they'll be able to support it then.

**Reply by w-r-z-k** [reply]
That's a good idea.  I opened a bug over there: https://github.com/LunarVim/LunarVim/issues/2087

Thanks

**Comment by w-r-z-k** [comment]
I swicthed to @ChristianChiarulli's [Neovim-from-scratch](https://github.com/LunarVim/Neovim-from-scratch) which I think is an implementation of his default [LunarVim](https://github.com/LunarVim) config expressed in a standardly configured fashion so people like me can mess with it using conventional nvim configuration knowledge.

Then I added in the sample `nvim-metals` config.  And the `*.worksheet.sc` was **not** recognized.

Then on a hunch, I commented out the line:

`--  use "neovim/nvim-lspconfig" -- enable LSP`

in the `.config/nvim/lua/plugins/lua` file.  Then the `*.worksheet.sc` **was** recognized and evaluated.  

So it looks like just having `nvim-lspconfig` there caused this problem.

I had understood that `nvim-lspconfig` could coexist with `nvim-metals` as long as `nvim-lspconfig` was not used to configure metals; which it wasn't, as least in so far I didn't do so myself.  But there must be something in there that overrides the filetype mappings or interferes somehow.  Not knowing how `nvim-lspconfig` or `nvim-metals` works, I can't really guess.

**Reply by ckipp01** [reply]
> I had understood that nvim-lspconfig could coexist with nvim-metals as long as nvim-lspconfig was not used to configure metals;

Correct. I have this locally in my config. I use `nivm-metals` for Scala, but `nvim-lspconfig` for other languages. I know others use this as well. I'm assuming that the thing you commented out is doing more than just enabling `nvim-lspconfig`, but also other stuff, which is probably what is conflicting.

**Reply by ckipp01** [reply]
One other thing to try out would be to try opening a `Test.sc` file. Does Metals start and recognize it as a Scala file?

**Reply by w-r-z-k** [reply]
Thanks!  I uncommented the line `use "neovim/nvim-lspconfig" -- enable LSP` then opened a `Test.sc` file and it asked if I wanted to import an ammonite script.  I said yes.  I entered the line `42` and it started to compile.  Exited and went into it again and it gave the message:

`LSP[metals] Error fetching Ammonite 2.3.8-124-2da846d2 for scala 2.13.7`

which didn't seem to affect my `42` script's compile.  Then I went into my existing `.worksheet.sc` file and it recognized and evaluated.  It did not work previously!  And I don't think I changed any configuration other than the uncommented line.

Somewhere along the line I saw Packer(?) pull down and install some patches, including for metals.  And I also saw an error message about something-_cmpe_ being installed?/referenced?/called? twice.  It flashes on the screen too quickly to capture in the mind.  Maybe it's in a log somewhere.

I find that nvim (independently of nvim-metals) bizarrely doesn't work then suddenly works.  I think it's a consequence of the various bootstrapping config mechanisms.

**Reply by ckipp01** [reply]
> And I also saw an error message about something-cmpe being installed?/referenced?/called? twice. It flashes on the screen too quickly to capture in the mind. Maybe it's in a log somewhere.

When this happens you should be able to do a `:messages` and see the message again.

> I find that nvim (independently of nvim-metals) bizarrely doesn't work then suddenly works. I think it's a consequence of the various bootstrapping config mechanisms.

I haven't personally experienced this and I actually run the nightly version of nvim. However, if you have a bunch of different type of configs all tied together then yea, there is a chance that stuff starts getting tripped up if it's not all organized.

**Reply by w-r-z-k** [reply]
With my latest minimal configuration, worksheets seem to be working just fine.  This must have been a configuration issue.  Thanks for the tip on `:messages` . Very useful!
